velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -751,6 +753,87 @@ TpchPlan TpchQueryBuilder::getQ14Plan() const {
+  std::vector<std::string> supplierColumns = {"s_suppkey", "s_comment"};
+  std::vector<std::string> partsuppColumns = {"ps_partkey", "ps_suppkey"};
+
+  const auto partSelectedRowType = getRowType(kPart, partColumns);
+  const auto& partFileColumns = getFileColumnNames(kPart);
+  const auto supplierSelectedRowType = getRowType(kSupplier, supplierColumns);
+  const auto& supplierFileColumns = getFileColumnNames(kSupplier);
+  const auto partsuppSelectedRowType = getRowType(kPartsupp, partsuppColumns);
+  const auto& partsuppFileColumns = getFileColumnNames(kPartsupp);
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId partScanNodeId;
+  core::PlanNodeId supplierScanNodeId;
+  core::PlanNodeId partsuppScanNodeId;
+
+  auto part = PlanBuilder(planNodeIdGenerator, pool_.get())
+                  .tableScan(
+                      kPart,
+                      partSelectedRowType,
+                      partFileColumns,
+                      {"p_size in (49, 14, 23, 45, 19, 3, 36, 9)"},
+                      "p_type NOT LIKE 'MEDIUM POLISHED%'")
+                  .capturePlanNodeId(partScanNodeId)
+                  // Neq is unsupported as a tableScan subfield filter for
+                  // Parquet source.
+                  .filter("p_brand <> 'Brand#45'")
+                  .planNode();
+
+  auto supplier = PlanBuilder(planNodeIdGenerator)
+                      .tableScan(
+                          kSupplier,
+                          supplierSelectedRowType,
+                          supplierFileColumns,
+                          {},
+                          "s_comment LIKE '%Customer%Complaints%'")
+                      .capturePlanNodeId(supplierScanNodeId)
+                      .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kPartsupp, partsuppSelectedRowType, partsuppFileColumns)
+          .capturePlanNodeId(partsuppScanNodeId)
+          .hashJoin(
+              {"ps_partkey"},
+              {"p_partkey"},
+              part,
+              "",
+              {"ps_suppkey", "p_brand", "p_type", "p_size"})
+          .hashJoin(
+              {"ps_suppkey"},
+              {"s_suppkey"},
+              supplier,
+              "",
+              {"ps_suppkey", "p_brand", "p_type", "p_size"},
+              core::JoinType::kAnti)
+          // Empty aggregate is used here to get the distinct count of
+          // ps_suppkey.
+          // approx_distinct could be used instead for getting the count of
+          // distinct ps_suppkey but since approx_distinct is non deterministic
+          // and the standard error can not be set to 0, it is not used here.
+          .singleAggregation({"p_brand", "p_type", "p_size", "ps_suppkey"}, {})
+          .partialAggregation(
+              {"p_brand", "p_type", "p_size"},
+              {"count(ps_suppkey) as supplier_cnt"})
+          .localPartition({})
+          .finalAggregation()
```

Comment:
I'm seeing that .singleAggregation({"p_brand", "p_type", "p_size", "ps_suppkey"} runs multi-threaded on a data not pre-partitioned on group-by keys. Is this intentional? It seems we need to ensure that we get globally distinct values before aggregating further.

Reply:
No, this was not the intention, the single aggregation was added simply to get the distinct count of ps_suppkey. Shall I partition the date on grouping keys before the single aggregation?

Reply:
We have to pre-partition on the `ps_suppkey` key I guess for the singleAggregation output to be globally distinct.

Reply:
Thanks, modified accordingly